### ISSUE TITLE
Avoid using is { ... } pattern when value is considered not null

### ISSUE THREAD

**Issue by drieseng** [meziantou/Meziantou.Analyzer#999]
While doing code reviews, I regularly come across needless use of the `is { ... }` pattern when the value is "guaranteed" to be not null.

Doing so adds an extra null check.

Example:
```c#
#nullable enable

private static void Do1(string value)
{
    // NOK; can be removed altogether since value is "guaranteed" to be not null; in a public API one should throw an ANE
    if (value is { })
    {
    }

    // NOK, can be simplified to `value.Length: > 0`
    if (value is { Length: > 0 })
    {
    }
}

#nullable disable

private static void Do2(string value)
{
    // OK; value can be null here
    if (value is { })
    {
    }

    // NOK, can be simplified to `value.Length: > 0`
    if (value is { Length: > 0 })
    {
    }
}

private static void Do3(string value)
{
    // OK, can be null here
    if (value is { Length: > 0 })
    {
    }
}
```

Would be nice if there was a rule to flag this.

**Comment by drieseng** [comment]
If you consider it feasible and useful to implement, this should probably go in the Performance category.

**Comment by meziantou** [comment]
I'm not sure if this is easy to do and I think this is a rabbit hole, so I'm not sure if I want to invest in this one.

Why not detecting useless `?.` or `is not null`?

**Comment by drieseng** [comment]
@meziantou: you're right, I'll submit separate issues for ... just kidding :-)

I'll close this issue. ReSharper code inspections already flag the first case.
You've already gone miles beyond what I could've asked of you.
Too bad there's so much red tape to cross in my company to get you compensated :(
